tests/record.cpp

```diff
@@ -0,0 +1,553 @@
+    }
+
+    jit_log(LogLevel::Info, "Replay:");
+    {
+        UInt32 i0(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+        UInt32 r0(2, 3, 4, 5, 6, 7, 8, 9, 10, 11);
+
+        uint32_t inputs[] = {
+            i0.index(),
+        };
+        uint32_t outputs[1];
+
+        jit_freeze_replay(recording, inputs, outputs);
+
+        jit_log(LogLevel::Info, "o0: %s", jit_var_str(outputs[0]));
+        jit_assert(jit_var_all(jit_var_eq(r0.index(), outputs[0])));
+    }
+
+    jit_freeze_destroy(recording);
+}
+
+/**
+ * This tests, weather it is possible to record multiple kernels in parallel.
+ * The variables of the kernels are of different size, therefore two kernels are
+ * generated. At replay these can be executed in parallel (LLVM) or sequence
+ * (CUDA).
+ */
+TEST_BOTH(06_parallel_kernels) {
+    Recording *recording;
+
+    jit_log(LogLevel::Info, "Recording:");
+    {
+        UInt32 i0(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+        UInt32 i1(0, 1, 2, 3, 4, 5);
+        UInt32 r0(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+        UInt32 r1(1, 2, 3, 4, 5, 6);
+
+        uint32_t inputs[] = {
+            i0.index(),
+            i1.index(),
+        };
+
+        jit_freeze_start(Backend, inputs, 2);
+
+        UInt32 o0 = i0 + 1;
+        UInt32 o1 = i1 + 1;
+        o0.schedule();
+        o1.schedule();
+        jit_eval();
+
+        uint32_t outputs[] = {
+            o0.index(),
+            o1.index(),
+        };
+
+        recording = jit_freeze_stop(Backend, outputs, 2);
+
+        jit_log(LogLevel::Info, "o0: %s", jit_var_str(outputs[0]));
+        jit_log(LogLevel::Info, "o1: %s", jit_var_str(outputs[1]));
+        jit_assert(jit_var_all(jit_var_eq(r0.index(), outputs[0])));
+        jit_assert(jit_var_all(jit_var_eq(r1.index(), outputs[1])));
+    }
+
+    jit_log(LogLevel::Info, "Replay:");
+    {
+        UInt32 i0(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+        UInt32 i1(1, 2, 3, 4, 5, 6);
+        UInt32 r0(2, 3, 4, 5, 6, 7, 8, 9, 10, 11);
+        UInt32 r1(2, 3, 4, 5, 6, 7);
+
+        uint32_t inputs[] = {
+            i0.index(),
+            i1.index(),
+        };
+        uint32_t outputs[2];
+
+        jit_freeze_replay(recording, inputs, outputs);
+
+        jit_log(LogLevel::Info, "o0: %s", jit_var_str(outputs[0]));
+        jit_log(LogLevel::Info, "o1: %s", jit_var_str(outputs[1]));
+        jit_assert(jit_var_all(jit_var_eq(r0.index(), outputs[0])));
+        jit_assert(jit_var_all(jit_var_eq(r1.index(), outputs[1])));
+    }
+
+    jit_freeze_destroy(recording);
+}
+
+/**
+ * This tests recording and replay of a horizontal reduction operation (hsum).
+ */
+TEST_BOTH(07_reduce_hsum) {
+    Recording *recording;
+
+    jit_log(LogLevel::Info, "Recording:");
+    {
+        UInt32 i0(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+        UInt32 r0 = opaque<UInt32>(45, 1);
+
+        uint32_t inputs[] = {
+            i0.index(),
+        };
+
+        jit_freeze_start(Backend, inputs, 1);
+
+        UInt32 o0 = hsum(i0);
```

Comment:
It would be more interesting if there was one operation creating a first kernel, and then `hsum()` is applied on the output of the first kernel.